cura/Settings/MachineManager.py

```diff
@@ -1496,3 +783,268 @@ def setSettingForAllExtruders(self, setting_name: str, property_name: str, prope
+            quality_changes_container = quality_changes_group.node_for_global.getContainer()
+        if quality_group.node_for_global:
+            quality_container = quality_group.node_for_global.getContainer()
+
+        self._global_container_stack.quality = quality_container
+        self._global_container_stack.qualityChanges = quality_changes_container
+
+        for position, extruder in self._global_container_stack.extruders.items():
+            quality_changes_node = quality_changes_group.nodes_for_extruders.get(position)
+            quality_node = quality_group.nodes_for_extruders.get(position)
+
+            quality_changes_container = self._empty_quality_changes_container
+            quality_container = self._empty_quality_changes_container
+            if quality_changes_node:
+                quality_changes_container = quality_changes_node.getContainer()
+            if quality_node:
+                quality_container = quality_node.getContainer()
+
+            extruder.quality = quality_container
+            extruder.qualityChanges = quality_changes_container
+
+        self._current_quality_group = quality_group
+        self._current_quality_changes_group = quality_changes_group
+        self.activeQualityGroupChanged.emit()
+        self.activeQualityChangesGroupChanged.emit()
+
+    def _setVariantNode(self, position, container_node):
+        self._global_container_stack.extruders[position].variant = container_node.getContainer()
+        self.activeVariantChanged.emit()
+
+    def _setGlobalVariant(self, container_node):
+        self._global_container_stack.variant = container_node.getContainer()
+
+    def _setMaterial(self, position, container_node = None):
+        if container_node:
+            self._global_container_stack.extruders[position].material = container_node.getContainer()
+        else:
+            self._global_container_stack.extruders[position].material = self._empty_material_container
+        # The _current_root_material_id is used in the MaterialMenu to see which material is selected
+        root_material_id = container_node.metadata["base_file"]
+        root_material_name = container_node.getContainer().getName()
+        if root_material_id != self._current_root_material_id[position]:
+            self._current_root_material_id[position] = root_material_id
+            self._current_root_material_name[position] = root_material_name
+            self.rootMaterialChanged.emit()
+
+    def activeMaterialsCompatible(self):
+        # check material - variant compatibility
+        if Util.parseBool(self._global_container_stack.getMetaDataEntry("has_materials", False)):
+            for position, extruder in self._global_container_stack.extruders.items():
+                if not extruder.material.getMetaDataEntry("compatible"):
+                    return False
+        return True
+
+    ## Update current quality type and machine after setting material
+    def _updateQualityWithMaterial(self):
```

Comment:
Something for the quality manager?

Reply:
No, this logic is explicitly for the active machine. This function is kinda like the old `setActiveQuality()` but way simpler.